File: leetcode/214.java
public class Solution {
    public String shortestPalindrome(String s) {
        if(s.length() <= 1) {
            return s;
        }
        // find the longest palindrome prefix
        int high = s.length() - 1;
        for(; high > 0 && !isPalindrome(s, 0, high); --high);
        String result = new StringBuilder(s.substring(high+1, s.length())).reverse().toString();
        return result + s;
    }

    boolean isPalindrome(String s, int low, int high) {
        for(; low < high; ++low, --high) {
            if(s.charAt(low) != s.charAt(high)) {
                return false;
            }
        }
        return true;
    }
}

File: leetcode/214.cpp
class Solution {
public:
    string shortestPalindrome(string s) {
        int suffixIndex = 0;
        for(int i=s.length() - 1; i>=0; --i) {
            if(s[i] == s[suffixIndex]) {
                ++suffixIndex;
            }
        }
        if(suffixIndex == s.length()) {
            return s;
        }
        int sublength = s.length() - suffixIndex;
        string prefix(sublength, 'a');
        for(int i=0; i<sublength; ++i) {
            prefix[i] = s[s.length() - i - 1];
        }
        return prefix + shortestPalindrome(s.substr(0, suffixIndex)) + s.substr(suffixIndex);
    }
};
